core/citrus-base/src/main/java/org/citrusframework/DefaultTestCase.java

```diff
@@ -289,7 +322,25 @@ private void gracefullyStopTimer() {
      * Run final test actions.
      */
     private void executeFinalActions(TestContext context) {
+        if (!finalActions.isEmpty() || !context.getFinalActions().isEmpty()) {
```

Comment:
I am not sure if this `if` is worth it. in the end, you'll safe no time. when entering the method you'll just end up with an empty loop, right? on the other hand, this method here looks more complicated because of it. might be worth a second guess.

Reply:
The thing is, that you then get the notifications although there are no final actions. I think this is a design issue. We decided not to send messages for a specific lifecycle if it is not in use.

Reply:
aah, sorry. it's about final actions vs final action reporters. now I got it. thx 😄